package-res/resources/web/pentaho/data/Filter.js

```diff
@@ -0,0 +1,89 @@
+/*!
+ * Copyright 2010 - 2016 Pentaho Corporation.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+define([
+  "./filter/IsEqual",
+  "./filter/IsIn",
+  "./filter/And",
+  "./filter/Or",
+  "./filter/Not"
+], function(IsEqual, IsIn, AndFilter, OrFilter, NotFilter) {
+  "use strict";
+
+
+  var RootFilter = OrFilter.extend({
+    constructor: function(filterSpec) {
+      if(filterSpec)
+        this.base(fromSpec(filterSpec));
+      else
+        this.base();
+    }
+  });
+
+  return {
+    // Leaf nodes
+    IsEqual: IsEqual,
+    IsIn: IsIn,
+    // Non-leaf nodes
+    Or: OrFilter,
+    And: AndFilter,
+    Not: NotFilter,
+    //Root: RootFilter,
+
+    /**
+     * Create a filter from a spec
+     * @param {Object} spec - Specification of a Filter
+     * @returns {*}
+     */
+    create: function (spec) {
+      return new RootFilter(spec);
+    }
+  };
+
+
+
+  function fromSpec(filterSpec) {
```

Comment:
This code seems to have several flaws. What's the spec syntax? I could use one or two snippets in a comment to be sure.
1. `registeredFilters` could be declared once, outside the function scope.
2. The outer loop accepts **in the same object** mixed keys with `$` and not. It would happily deserialize `{foo: "1", $equal: 2}` although this is a-syntactic. You should have to functions, one for each expected syntactic context.
3. `$not` ends up being handled specially, which given we're using a dictionary lookup for constructors is weird. We could have a dictionary lookup for factories.
4. `for(operator ...` is strange cause you return in the first iteration. I see the saving of an `if` test and of an auxiliary variable, but it is confusing.
5. `filterSpec[property]` — variables exist.
6. no error handling.